---
title: "GoSeq_Ojo"
author: "Veronica Radice"
date: "16/08/2021"
output: html_document
---

Gene Ontology testing for RNA-seq datasets
- accounts for the gene length bias in detection of over-representation


```{r}
# for first time installation
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("goseq")

library(goseq)
library(dplyr)
```

# Porites astreoides - Host

### Annotation
```{r}
past.annotation <- read.csv("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/GoSeq_OA/Past_hybridreference_totalannotated_Final.csv")

past.annotation <- past.annotation[,2:26]

# select ContigName column
assayed.genes <- past.annotation[,1]

past.annotation[sapply(past.annotation, is.character)] <- lapply(past.annotation[sapply(past.annotation, is.character)], as.factor)

head(past.annotation)
```

```{r}
head(levels(past.annotation$GO))
```

```{r}
goterms <- strsplit(as.character(past.annotation$GO), split=" // ")
names(goterms) <- past.annotation$gene
```


#### import Past seq2iso table
Carly Kenkel's P. astreoides transcriptome reference
```{r}
past_seq2iso <- read.table("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/GoSeq_OA/past_seq2iso_suffixed.tab", header = FALSE, sep = "\t", stringsAsFactors=T)

past_seq2iso <- past_seq2iso %>% 
  dplyr::rename(isotig = V1,
         gene = V2)

head(past_seq2iso)
```


### Counts
```{r}
# All genes
#past.counts.all <- read.csv("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/Porites_astreoides/host/counts_normalized_Past_Host.csv")
```


# Ojo-Ojo vs. Lagoon-Ojo 
```{r}
# Absolute number of DEG adjusted p-value < 0.1
past.DE.genes.OO.LO <- readRDS("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/GoSeq_OA/Past_Host_DE_genes_OO_LO_padj.1.rds")

past.DE.genes.OO.LO <- as.data.frame(past.DE.genes.OO.LO)

# add group pairwise contrast label
#past.DE.genes.OO.LO$group <- 'OO-LO'

# Log Fold Change - Up- vs. Down-regulated
#past.DE.genes.OO.LO$LFC <- ifelse(past.DE.genes.OO.LO$log2FoldChange>0, 'Up', 'Down')

# make all character variables as factors
#past.DE.genes.OO.LO[sapply(past.DE.genes.OO.LO, is.character)] <- lapply(past.DE.genes.OO.LO[sapply(past.DE.genes.OO.LO, is.character)], as.factor)
```


```{r}
past.DE.genes.OO.LO <- right_join(past_seq2iso, past.DE.genes.OO.LO, by = c("gene"))

rownames(past.DE.genes.OO.LO) <- past.DE.genes.OO.LO[,1]

# ends up removing column
# past.DE.genes.OO.LO <- past.DE.genes.OO.LO %>% 
#   column_to_rownames(var="gene")

# count the number of significantly differentially expressed genes
sum(past.DE.genes.OO.LO$padj < 0.1, na.rm =T)
sum(past.DE.genes.OO.LO$log2FoldChange!=0)
```

## Differentially expressed genes
```{r}
# each element of your vector should be either a 1 or a 0, 
# where 1 indicates that the gene is differentially expressed and 0 that it is not.
past.DE.genes.OO.LO$DEG <- '1'

de.genes <- full_join(past.DE.genes.OO.LO, past.annotation, by = c("gene", "isotig"))

de.genes <- de.genes %>%
    mutate(DEG = if_else(is.na(DEG), '0', DEG),
           padj=ifelse(is.na(padj), 1, padj),
           log2FoldChange=ifelse(is.na(log2FoldChange), 0, log2FoldChange))

rownames(de.genes) <- de.genes[,10]

# replaces all NA with a number
# de.genes <- de.genes %>% 
#   mutate_if(is.numeric, replace_na, replace = 0.9)

# to check for any rows with NA in ContigName column
#de.genes[duplicated(de.genes$ContigName),]
#anyNA(de.genes$ContigName)
```


In order to perform a GO analysis of your RNA-seq data, goseq only requires a simple named
vector, which contains two pieces of information.
1. Measured genes: all genes for which RNA-seq data was gathered for your experiment. Each
element of your vector should be named by a unique gene identifier.
2. Differentially expressed genes: each element of your vector should be either a 1 or a
0, where 1 indicates that the gene is differentially expressed and 0 that it is not.


### DE genes as integer vector
```{r}
genes <- as.integer(de.genes$padj<.1 & de.genes$log2FoldChange!=0)
names(genes)<-row.names(de.genes)
table(genes)
```


## Fit the Probability Weighting Function (PWF)

*nullp expects a vector for bias.data, not a data.frame*

The output of nullp contains all the data used to create the PWF, as well as the PWF itself.
It is a data frame with 3 columns, named "DEgenes", "bias.data" and "pwf" with the rownames set to the gene names.

```{r}
pwf <- nullp(genes, bias.data = past.annotation[past.annotation$ContigName%in%names(genes),"ContigLength"])
```

```{r}
plotPWF(pwf = pwf, binsize = 500)
```


Selection-unbiased testing for category enrichment amongst differentially expressed (DE) genes

```{r}
GOGOGO <- goseq(pwf, gene2cat = goterms, use_genes_without_cat=FALSE, method = "Wallenius")
GOGOGO$padj <- p.adjust(GOGOGO$over_represented_pvalue, method = "fdr")
```

When default applied (use_genes_without_cat=FALSE):

*"For 22722 genes, we could not find any categories. These genes will be excluded.*
To force their use, please run with use_genes_without_cat=TRUE (see documentation).
This was the default behavior for version 1.15.1 and earlier."


```{r}
sum(past.annotation$GO=="No_GOcodes")
sum(past.annotation$GO=="No_GO")
sum(past.annotation$GO=="No_GOcodes")+sum(past.annotation$GO=="No_GO")
```

```{r}
sum(grepl("No", past.annotation$GO))
```

*22722 genes is exact number of isotigs that have corresponding (but sometimes the same/overlapping) isogroups*
```{r}
sum(grepl("isotig", past.annotation$ContigName))
```

```{r}
head(rownames(pwf))
```

*Gene names from DESeq analysis based on isogroups NOT isotigs*


```{r}
GOGOGO <- goseq(pwf, gene2cat = goterms, use_genes_without_cat=TRUE, method = "Wallenius")
GOGOGO$padj <- p.adjust(GOGOGO$over_represented_pvalue, method = "fdr")
head(GOGOGO)
```


```{r}
sum(GOGOGO$padj<0.1, na.rm=T)
GOGOGO[GOGOGO$padj<0.1,]
```

```{r}
enriched.GO <- GOGOGO$category[p.adjust(GOGOGO$over_represented_pvalue, method="BH")<.1]
head(enriched.GO)
```


########################################################################

# Ojo-Control vs. Lagoon-Control
```{r}
# Absolute number of DEG adjusted p-value < 0.1
past.DE.genes.OC.LC <- readRDS("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/GoSeq_OA/Past_Host_DE_genes_OC_LC_padj.1.rds")

past.DE.genes.OC.LC <- as.data.frame(past.DE.genes.OC.LC)

past.DE.genes.OC.LC <- right_join(past_seq2iso, past.DE.genes.OC.LC, by = c("gene"))

rownames(past.DE.genes.OC.LC) <- past.DE.genes.OC.LC[,1]

# count the number of significantly differentially expressed genes
#sum(past.DE.genes.OC.LC$padj < 0.1, na.rm =T)
#sum(past.DE.genes.OC.LC$log2FoldChange!=0)

past.DE.genes.OC.LC$DEG <- '1'

de.genes <- full_join(past.DE.genes.OC.LC, past.annotation, by = c("gene", "isotig"))

de.genes <- de.genes %>%
    mutate(DEG = if_else(is.na(DEG), '0', DEG),
           padj=ifelse(is.na(padj), 1, padj),
           log2FoldChange=ifelse(is.na(log2FoldChange), 0, log2FoldChange))

rownames(de.genes) <- de.genes[,10]

genes <- as.integer(de.genes$padj<.1 & de.genes$log2FoldChange!=0)
names(genes)<-row.names(de.genes)
table(genes)
```

```{r}
pwf <- nullp(genes, bias.data = past.annotation[past.annotation$ContigName%in%names(genes),"ContigLength"])
```

```{r}
GOGOGO <- goseq(pwf, gene2cat = goterms, use_genes_without_cat=TRUE, method = "Wallenius")
GOGOGO$padj <- p.adjust(GOGOGO$over_represented_pvalue, method = "fdr")
head(GOGOGO)
```

```{r}
sum(GOGOGO$padj<0.1, na.rm=T)
GOGOGO[GOGOGO$padj<0.1,]
```

```{r}
enriched.GO <- GOGOGO$category[p.adjust(GOGOGO$over_represented_pvalue, method="BH")<.1]
head(enriched.GO)
```


########################################################################

# Ojo-Ojo vs Ojo-Control
```{r}
# Absolute number of DEG adjusted p-value < 0.1
past.DE.genes.OO.OC <- readRDS("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/GoSeq_OA/Past_Host_DE_genes_OO_OC_padj.1.rds")

past.DE.genes.OO.OC <- as.data.frame(past.DE.genes.OO.OC)

past.DE.genes.OO.OC <- right_join(past_seq2iso, past.DE.genes.OO.OC, by = c("gene"))

rownames(past.DE.genes.OO.OC) <- past.DE.genes.OO.OC[,1]

# count the number of significantly differentially expressed genes
#sum(past.DE.genes.OC.LC$padj < 0.1, na.rm =T)
#sum(past.DE.genes.OC.LC$log2FoldChange!=0)

past.DE.genes.OO.OC$DEG <- '1'

de.genes <- full_join(past.DE.genes.OO.OC, past.annotation, by = c("gene", "isotig"))

de.genes <- de.genes %>%
    mutate(DEG = if_else(is.na(DEG), '0', DEG),
           padj=ifelse(is.na(padj), 1, padj),
           log2FoldChange=ifelse(is.na(log2FoldChange), 0, log2FoldChange))

rownames(de.genes) <- de.genes[,10]

genes <- as.integer(de.genes$padj<.1 & de.genes$log2FoldChange!=0)
names(genes)<-row.names(de.genes)
table(genes)
```

```{r}
pwf <- nullp(genes, bias.data = past.annotation[past.annotation$ContigName%in%names(genes),"ContigLength"])
```

```{r}
GOGOGO <- goseq(pwf, gene2cat = goterms, use_genes_without_cat=TRUE, method = "Wallenius")
GOGOGO$padj <- p.adjust(GOGOGO$over_represented_pvalue, method = "fdr")
head(GOGOGO)
```

```{r}
GOGOGO.cat.terms <- GOGOGO[,c(1,6,7)]
#write.csv("GO_Past_category_terms_list.csv", x = GOGOGO.cat.terms)
```


```{r}
sum(GOGOGO$padj<0.1, na.rm=T)
GOGOGO[GOGOGO$padj<0.1,]
```

```{r}
enriched.GO <- GOGOGO$category[p.adjust(GOGOGO$over_represented_pvalue, method="BH")<.1]
head(enriched.GO)
```


########################################################################

# Lagoon-Ojo vs. Lagoon-Control
```{r}
# Absolute number of DEG adjusted p-value < 0.1
past.DE.genes.LO.LC <- readRDS("~/Documents/Rprojects/postdoc\ Rprojects/ODU_postdoc_Rprojects/Paper_Ojo_gene-expression/GoSeq_OA/Past_Host_DE_genes_LO_LC_padj.1.rds")

past.DE.genes.LO.LC <- as.data.frame(past.DE.genes.LO.LC)

past.DE.genes.LO.LC <- right_join(past_seq2iso, past.DE.genes.LO.LC, by = c("gene"))

rownames(past.DE.genes.LO.LC) <- past.DE.genes.LO.LC[,1]

# count the number of significantly differentially expressed genes
#sum(past.DE.genes.OC.LC$padj < 0.1, na.rm =T)
#sum(past.DE.genes.OC.LC$log2FoldChange!=0)

past.DE.genes.LO.LC$DEG <- '1'

de.genes <- full_join(past.DE.genes.LO.LC, past.annotation, by = c("gene", "isotig"))

de.genes <- de.genes %>%
    mutate(DEG = if_else(is.na(DEG), '0', DEG),
           padj=ifelse(is.na(padj), 1, padj),
           log2FoldChange=ifelse(is.na(log2FoldChange), 0, log2FoldChange))

rownames(de.genes) <- de.genes[,10]

genes <- as.integer(de.genes$padj<.1 & de.genes$log2FoldChange!=0)
names(genes)<-row.names(de.genes)
table(genes)
```

```{r}
pwf <- nullp(genes, bias.data = past.annotation[past.annotation$ContigName%in%names(genes),"ContigLength"])
```

```{r}
GOGOGO <- goseq(pwf, gene2cat = goterms, use_genes_without_cat=TRUE, method = "Wallenius")
GOGOGO$padj <- p.adjust(GOGOGO$over_represented_pvalue, method = "fdr")
head(GOGOGO)
```

```{r}
sum(GOGOGO$padj<0.1, na.rm=T)
GOGOGO[GOGOGO$padj<0.1,]
```

```{r}
enriched.GO <- GOGOGO$category[p.adjust(GOGOGO$over_represented_pvalue, method="BH")<.1]
head(enriched.GO)
```



########################################################################


*Did not try / run any of the below because no enriched GO terms*


Plot top 10 results
```{r}
goResults %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count")
```


GO_explained
```{r}
library(GO.db)
GOTERM[[goResults$category[1]]]

# library(GO.db)
# for(go in enriched.GO[1:10]){
# 	print(GOTERM[[go]])
# 	cat("--------------------------------------\n")
# }
```


KEGG pathway enrichment analysis
```{r}
library(clusterProfiler)
search_kegg_organism('mmu', by='kegg_code')
```

```{r}
sigGenes <- shrinkLvV$Entrez[ shrinkLvV$FDR < 0.01 & 
                              !is.na(shrinkLvV$FDR) &
                              abs(shrinkLvV$logFC) > 1 ]
sigGenes <- na.exclude(sigGenes)
kk <- enrichKEGG(gene = sigGenes, organism = 'mmu')
head(kk, n=10)
```


Visualise a pathway
view the Kegg pathway with the genes colours in in your browser
```{r}
browseKEGG(kk, 'mmu03320')
```

Visualise a pathway
as a file
```{r}
library(pathview)
logFC <- annotLvV$logFC
names(logFC) <- annotLvV$Entrez
pathview(gene.data = logFC, 
         pathway.id = "mmu03320", 
         species = "mmu", 
         limit = list(gene=5, cpd=1))
```


KEGG_mappings (eval = FALSE)
```{r}
# Get the mapping from ENSEMBL 2 Entrez
en2eg=as.list(org.Hs.egENSEMBL2EG)

# Get the mapping from Entrez 2 KEGG
eg2kegg=as.list(org.Hs.egPATH)

# Define a function which gets all unique KEGG IDs
# associated with a set of Entrez IDs
grepKEGG=function(id,mapkeys){unique(unlist(mapkeys[id],use.names=FALSE))}

# Apply this function to every entry in the mapping from
# ENSEMBL 2 Entrez to combine the two maps
kegg=lapply(en2eg,grepKEGG,eg2kegg)
head(kegg)

# KEGG (eval = FALSE)
pwf=nullp(genes,"hg19","ensGene")
KEGG=goseq(pwf,gene2cat=kegg)
head(KEGG)

#KEGG_goseq
pwf=nullp(genes,'hg19','ensGene')
KEGG=goseq(pwf,'hg19','ensGene',test.cats="KEGG")
head(KEGG)

# KEGG_from_db
kegg=as.list(org.Hs.egPATH)
head(kegg)
```




### Session Info
```{r}
sessionInfo()
```

